### PR TITLE
feat(mito2): time series memtable

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR implements a `TimeSeriesMemtable` that group rows according to their primary key.
- ensures memtable will not waste memory to store duplicated primary keys
- each time series (primary key) will have an `active` part and several `frozen` parts
  - `active` part is for writing, newly written rows will be pushed into the `active` part
  - `frozen` parts are for reading. When scanning the time series, the `active` part will be frozen and all `frozen` parts will be compacted to one single `Value` (containing all the field vectors including sequence and op_type) so that it can be expanded to a `Batch` struct.
    - Maybe we can postpone the `compact` operation when values are small so that we won't waste too much CPU cycles on memtable compaction. But it's not been implemented in this PR.

This PR does not:
- apply projection when yielding batches
- implement the memory footprint tracker

## Checklist

- [X]  I have written the necessary rustdoc comments.
- [X]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/storage/src/metadata.rs [hunk 1]
```diff
@@ -483,6 +483,12 @@ impl ColumnsMetadata {
         self.columns.iter().take(self.row_key_end)
     }
 
+    /// Returns an iterator to all primary key columns.
+    /// Primary key columns do not contain timestamp columns.
+    pub fn iter_primary_key_columns(&self) -> impl Iterator<Item = &ColumnMetadata> {
+        self.columns.iter().take(self.timestamp_key_index)
+    }
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/storage/src/metadata.rs]: Looks like this method is not used anywhere.

File: src/mito2/src/memtable/time_series.rs [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+    pub initial_builder_capacity: usize,
+}
+
+impl TimeSeriesMemtableConfig {
+    fn validate(&self) -> Result<()> {
+        ensure!(
+            self.bucket_num < 256,
+            InvalidMemtableConfigSnafu {
+                config: format!("{:?}", self),
+            }
```
Comment (waynexia) [inline src/mito2/src/memtable/time_series.rs:56]: sugg: add a `reason` field like "too many buckets"

**Review by waynexia — APPROVED** [review]

File: src/mito2/src/memtable/time_series.rs [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+        );
+        Ok(())
+    }
+}
+
+impl Default for TimeSeriesMemtableConfig {
+    fn default() -> Self {
+        Self {
+            bucket_num: 16,
+            initial_builder_capacity: 128,
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:66]: `128` may be too large. For an int64, 128 elements require at least 1k bytes. Consider smaller values like 8 and 16.

File: src/mito2/src/memtable/time_series.rs [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+            buckets,
+        }
+    }
+}
+
+impl SeriesSet {
+    /// Returns [Series] at given bucket and index.
+    fn get_bucket_by_index(&self, idx: &SeriesIndex) -> Option<Arc<RwLock<Series>>> {
+        debug_assert!((idx.0 as usize) < self.buckets.len());
+        self.buckets[idx.0 as usize].get_series(idx.1)
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:188]: `buckets[idx]` already does the assertion.
```suggestion
        self.buckets[idx.0 as usize].get_series(idx.1)
```

File: src/mito2/src/memtable/time_series.rs [hunk 5]
```diff
@@ -0,0 +1,840 @@
...
+        let mut series = self.series.write().unwrap();
+        let idx = series.len();
+        series.push(s);
+        idx
+    }
+}
+
+/// A `Series` holds a list of field values of some given primary key.
+struct Series {
+    initial_builder_capacity: usize,
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:300]: Could we define this as a constant or pass it at each function call? Storing this at each series is a little wasteful. Same as the RegionMetadataRef.

File: src/mito2/src/memtable/time_series.rs [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+        let mut arrays = Vec::with_capacity(3 + self.fields.len());
+        arrays.push(self.timestamp.to_arrow_array());
+        arrays.push(self.sequence.to_arrow_array());
+        arrays.push(self.op_type.to_arrow_array());
+        arrays.extend(self.fields.iter().map(|f| f.to_arrow_array()));
+
+        let fields = arrays
+            .iter()
+            .map(|arr| datatypes::arrow::row::SortField::new(arr.data_type().clone()))
+            .collect();
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:435]: I think we only need to sort by `timestamp` and `sequence`. Here it uses all columns to create rows for sort.

**Review by evenyag — APPROVED** [review]

**Pull request merged**